[PLAT-112] Add playlists metadata upload route and test

hareeshnagaraj opened the pull request:
### Description
<!--
What is the purpose of this PR?
What is the current behavior? New behavior?
Relevant links and/or information pertaining to PR?
-->
Add playlist table, playlist metadata upload route, db migration, integration test

Patching changes partially finished in this PR - https://github.com/AudiusProject/audius-protocol/pull/2977

### Tests
<!--
List any automated test coverage added, if current automated tests are not sufficient.
List the manual tests and repro instructions to verify that this PR works as anticipated.
Include log analysis if possible.
If this change impacts clients, make sure that you have tested the clients!
-->

Integration tests now passing, up/down migration validated manually  with and without state in existing tables + new Playlists table.

### Monitoring - How will this change be monitored? Are there sufficient logs / alerts?
No additional monitoring has been added for this PR but can be added in follow ups if necessary. As there is no anticipated effect to happy write path for uploads of metadata or images I don't believe it is required in this PR. Queries to playlist metadata objects should also provide clarity as this feature is developed


<!-- For features that are critical or could fail silently please describe the monitoring/alerting being added. -->


<!--
================ REMINDER: ================
If this PR touches a critical flow (such as Indexing, Uploads, Gateway or the Filesystem), make sure to add the `requires-special-attention` label.

** Add relevant labels as necessary. **
-->

hareeshnagaraj commented on creator-node/sequelize/migrations/20220427014326-create_playlist_table_and_clock_record_type.js:
```diff
@@ -0,0 +1,156 @@
+
+    // Create Playlist table
+    await createPlaylistTable(queryInterface, Sequelize, transaction)
+    await addPlaylistToClockRecordSourceTables(queryInterface, transaction)
```
remove unused transaction parameter as this cannot exist

SidSethi reviewed (commented):
good stuff - would def be good to test with state, esp with the non-standard enum migration

also could you fill out monitoring pr desc

looks great though

hareeshnagaraj commented on creator-node/sequelize/migrations/20220427014326-create-playlist-table-and-clock-record-type.js:
```diff
@@ -0,0 +1,164 @@
+  }
+}
+
+async function addPlaylistToClockRecordSourceTables(
```
great callout, will add more explanation here

hareeshnagaraj commented on creator-node/sequelize/migrations/20220427014326-create-playlist-table-and-clock-record-type.js:
```diff
@@ -0,0 +1,164 @@
+    )
+  } catch (e) {
+    console.log(e)
+  }
```
interesting, will confirm this

hareeshnagaraj commented:
@SidSethi - there's an issue with the down migration here in the case that a Playlist entry has already been added to ClockRecords. Further details below but tl;dr is that once an enum value has been inserted into a record (as in a ClockRecord with type = Playlist) , we can't remove that type entirely by renaming it. In my opinion we could either update all the `Playlist` values to something else or simply not modify the enum at all. 

### Verifying migration behavior with state:

Steps taken:

- Initialized local system
- Created user, added metadata and profile picture

User info:

```bash
audius_creator_node=# select * from "CNodeUsers" ;
 lastLogin |              walletPublicKey               |         createdAt          |         updatedAt          |            cnodeUserUUID             | latestBlockNumber | clock
-----------+--------------------------------------------+----------------------------+----------------------------+--------------------------------------+-------------------+-------
           | 0xee73894f48454c478789284d960486e27ab9bcc1 | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:35:54.951+00 | d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |              1483 |     9
```

Files info:

```bash
audius_creator_node=# select "cnodeUserUUID","multihash","clock","fileName","type" from "Files";
            cnodeUserUUID             |                   multihash                    | clock |   fileName    |   type
--------------------------------------+------------------------------------------------+-------+---------------+----------
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmX3uGhDU9a8sC5LRceH1LMemXwmQqfaLQVmTT4rM2tHtZ |     1 |               | metadata
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUwGD453zzANwGYgmH9cNjBbkWDXJYMZXKmGNKChXHws3 |     3 |               | dir
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmWgv9ZFTUXhaWEDvm7f8KQLrW9Xe11GFqP3SYA94Ptt1P |     4 | 150x150.jpg   | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmNydje7o5jmrJTxGEp8naviWsRjKwsXnJvVrZiiKYjewH |     5 | 480x480.jpg   | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmbbMKFxn7YKJGLVCPwFRmSxdQqyD1sMxxKF8Zei9Zanjr |     6 | 1000x1000.jpg | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUnyPBxMU3RKNB1f5xc8uMUYeJq3EuQ2zMnKtVijidrtW |     7 | original.jpg  | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUxeWCFAFCrBNniMzxDXSMRxzUNW8n65Qb7HggLnJiFGC |     8 |               | metadata
(7 rows)
```

Confirm latest DB migration w/playlists has been applied:

```bash
ubuntu@hn-remote-dev-2:~/audius-protocol/creator-node$ ./node_modules/.bin/sequelize db:migrate:status

Sequelize CLI [Node: 14.17.5, CLI: 5.5.1, ORM: 4.44.4]

Loaded configuration file "sequelize/sequelize-config.json".
Using environment "development".
sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:245:13
up 20181121184323-create-user.js
up 20181122035259-create-file.js
up 20181129193321-create-audius-user.js
up 20181204021851-create-session-token.js
up 20181207194524-create-track.js
up 20181211183951-create-track-file.js
up 20190711143402-add_uuid_to_users.js
up 20190805032224-add_on_update_cascade_to_session_tokens.js
up 20190827175430-add-types-to-files.js
up 20190906221141-add-blocknumber-to-cnodeuser.js
up 20191014204226-add-indexes.js
up 20191108022324-associate-files-with-trackuuid.js
up 20200623002237-add-parent-file-data.js
up 20200724154816-sync-indexes.js
up 20200904183206-add-track-uuid-index.js
up 20200911004845-allow-track-and-audiusUsers-appends.js
up 20200918150546-add-clock.js
up 20200922131913-post_vector_clock_db_migrations.js
up 20200928233308-create-content-blacklist.js
up 20201105015155-add-cid-and-update-content-blacklist.js
up 20210722020837-add-skipped-column-to-files-table.js
up 20210831162548-add-index-files-sourcefile.js
up 20220427014326-create-playlist-table-and-clock-record-type.js
```

Undo latest migration:

```bash
ubuntu@hn-remote-dev-2:~/audius-protocol/creator-node$ ./node_modules/.bin/sequelize db:migrate:undo --name 20220427014326-create-playlist-table-and-clock-record-type.js

Sequelize CLI [Node: 14.17.5, CLI: 5.5.1, ORM: 4.44.4]

Loaded configuration file "sequelize/sequelize-config.json".
Using environment "development".
sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:245:13
== 20220427014326-create-playlist-table-and-clock-record-type: reverting =======
== 20220427014326-create-playlist-table-and-clock-record-type: reverted (0.020s)
```

Next, updated cover photo and metadata without migration present:

```bash
audius_creator_node=# select * from "CNodeUsers" ;
 lastLogin |              walletPublicKey               |         createdAt          |         updatedAt          |            cnodeUserUUID             | latestBlockNumber | clock
-----------+--------------------------------------------+----------------------------+----------------------------+--------------------------------------+-------------------+-------
           | 0xee73894f48454c478789284d960486e27ab9bcc1 | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:49:46.996+00 | d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |              1649 |    15
(1 row)

audius_creator_node=# select "cnodeUserUUID","multihash","clock","fileName","type" from "Files";
            cnodeUserUUID             |                   multihash                    | clock |   fileName    |   type
--------------------------------------+------------------------------------------------+-------+---------------+----------
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmX3uGhDU9a8sC5LRceH1LMemXwmQqfaLQVmTT4rM2tHtZ |     1 |               | metadata
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUwGD453zzANwGYgmH9cNjBbkWDXJYMZXKmGNKChXHws3 |     3 |               | dir
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmWgv9ZFTUXhaWEDvm7f8KQLrW9Xe11GFqP3SYA94Ptt1P |     4 | 150x150.jpg   | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmNydje7o5jmrJTxGEp8naviWsRjKwsXnJvVrZiiKYjewH |     5 | 480x480.jpg   | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmbbMKFxn7YKJGLVCPwFRmSxdQqyD1sMxxKF8Zei9Zanjr |     6 | 1000x1000.jpg | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUnyPBxMU3RKNB1f5xc8uMUYeJq3EuQ2zMnKtVijidrtW |     7 | original.jpg  | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUxeWCFAFCrBNniMzxDXSMRxzUNW8n65Qb7HggLnJiFGC |     8 |               | metadata
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmfJmo3gLrZKUrRQuV1YSDt6FmP8M3KPETmAdUgAFBE94h |    10 |               | dir
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmWmEV7fh6TmGJKLuBoH3PiJ93PxJypADAxKWR8b9R1Hgb |    11 | 640x.jpg      | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmeWdQ9bjeNbWitgkgPnxddSShhxET2hWBRmsUucRtSP36 |    12 | 2000x.jpg     | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUhPPAWVNH3P5FMpQUUueKx81YL7Eu57wBsxLgmCSLbP6 |    13 | original.jpg  | image
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmXR2Lx1ceCDQ9otvTmjbYL552rEr1qrZHcc4AR1fXgF1m |    14 |               | metadata
(12 rows)
```

Re-running migration (upward testing w/state):

```bash
ubuntu@hn-remote-dev-2:~/audius-protocol/creator-node$ ./node_modules/.bin/sequelize db:migrate

Sequelize CLI [Node: 14.17.5, CLI: 5.5.1, ORM: 4.44.4]

Loaded configuration file "sequelize/sequelize-config.json".
Using environment "development".
sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:245:13
== 20220427014326-create-playlist-table-and-clock-record-type: migrating =======
== 20220427014326-create-playlist-table-and-clock-record-type: migrated (0.028s)
```

Inserting a record into Playlists table (following this - [https://github.com/AudiusProject/audius-protocol/blob/6fef5644f572934c5ce21e67d97682ce2ca5ac00/creator-node/src/dbManager.js#L273](https://github.com/AudiusProject/audius-protocol/blob/6fef5644f572934c5ce21e67d97682ce2ca5ac00/creator-node/src/dbManager.js#L273))

```bash
audius_creator_node=# update "CNodeUsers" set clock = clock + 1 where text("cnodeUserUUID") = 'd3f8e8ea-5e33-41bc-b004-8fc6ca4626cd';
UPDATE 1

audius_creator_node=# select * from "CNodeUsers";
 lastLogin |              walletPublicKey               |         createdAt          |         updatedAt          |            cnodeUserUUID             | latestBlockNumber | clock
-----------+--------------------------------------------+----------------------------+----------------------------+--------------------------------------+-------------------+-------
           | 0xee73894f48454c478789284d960486e27ab9bcc1 | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:49:46.996+00 | d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |              1649 |    16

INSERT INTO CLOCKRECORDS:

audius_creator_node=# INSERT INTO "ClockRecords" ("cnodeUserUUID", "clock", "sourceTable", "createdAt", "updatedAt") VALUES('d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd', 16, 'Playlist', '2022-06-22 21:29:13.755+00', '2022-06-22 21:29:13.755+00');

audius_creator_node=# select * from "ClockRecords";
            cnodeUserUUID             | clock | sourceTable |         createdAt          |         updatedAt
--------------------------------------+-------+-------------+----------------------------+----------------------------
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |     1 | File        | 2022-06-22 21:29:14.129+00 | 2022-06-22 21:29:14.129+00
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |     2 | AudiusUser  | 2022-06-22 21:29:18.675+00 | 2022-06-22 21:29:18.675+00
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |     3 | File        | 2022-06-22 21:35:45.041+00 | 2022-06-22 21:35:45.041+00
...
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    12 | File        | 2022-06-22 21:49:38.927+00 | 2022-06-22 21:49:38.927+00
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    13 | File        | 2022-06-22 21:49:38.934+00 | 2022-06-22 21:49:38.934+00
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    14 | File        | 2022-06-22 21:49:39.721+00 | 2022-06-22 21:49:39.721+00
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    15 | AudiusUser  | 2022-06-22 21:49:46.984+00 | 2022-06-22 21:49:46.984+00
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    16 | Playlist    | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:29:13.755+00

INSERT INTO PLAYLISTS:
audius_creator_node=# INSERT INTO "Playlists" ("cnodeUserUUID", "metadataFileUUID","metadataJSON","blockchainId","clock","createdAt","updatedAt") VALUES('d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd','1a3ee179-6e89-473f-924c-36fc6089bda6','{}',1000,16,'2022-06-22 21:29:13.755+00','2022-06-22 21:29:13.755+00');

audius_creator_node=# select * from "Playlists";
            cnodeUserUUID             | clock |           metadataFileUUID           | metadataJSON | blockchainId | playlistImageFileUUID |         createdAt          |         updatedAt
--------------------------------------+-------+--------------------------------------+--------------+--------------+-----------------------+----------------------------+----------------------------
 d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    16 | 1a3ee179-6e89-473f-924c-36fc6089bda6 | {}           |         1000 |                       | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:29:13.755+00
```

Downgrading migration:

```bash
== 20220427014326-create-playlist-table-and-clock-record-type: reverting =======
DatabaseError [SequelizeDatabaseError]: invalid input value for enum "enum_ClockRecords_sourceTable": "Playlist"
    at Query.formatError (/home/ubuntu/audius-protocol/creator-node/node_modules/sequelize/lib/dialects/postgres/query.js:363:16)
    at /home/ubuntu/audius-protocol/creator-node/node_modules/sequelize/lib/dialects/postgres/query.js:86:18
    at tryCatcher (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:725:18)
    at _drainQueueStep (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/async.js:93:12)
```

SidSethi commented:
> @SidSethi - there's an issue with the down migration here in the case that a Playlist entry has already been added to ClockRecords. Further details below but tl;dr is that once an enum value has been inserted into a record (as in a ClockRecord with type = Playlist) , we can't remove that type entirely by renaming it. In my opinion we could either update all the `Playlist` values to something else or simply not modify the enum at all.
> 
> ### Verifying migration behavior with state:
> Steps taken:
> 
> * Initialized local system
> * Created user, added metadata and profile picture
> 
> User info:
> 
> ```shell
> audius_creator_node=# select * from "CNodeUsers" ;
>  lastLogin |              walletPublicKey               |         createdAt          |         updatedAt          |            cnodeUserUUID             | latestBlockNumber | clock
> -----------+--------------------------------------------+----------------------------+----------------------------+--------------------------------------+-------------------+-------
>            | 0xee73894f48454c478789284d960486e27ab9bcc1 | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:35:54.951+00 | d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |              1483 |     9
> ```
> 
> Files info:
> 
> ```shell
> audius_creator_node=# select "cnodeUserUUID","multihash","clock","fileName","type" from "Files";
>             cnodeUserUUID             |                   multihash                    | clock |   fileName    |   type
> --------------------------------------+------------------------------------------------+-------+---------------+----------
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmX3uGhDU9a8sC5LRceH1LMemXwmQqfaLQVmTT4rM2tHtZ |     1 |               | metadata
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUwGD453zzANwGYgmH9cNjBbkWDXJYMZXKmGNKChXHws3 |     3 |               | dir
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmWgv9ZFTUXhaWEDvm7f8KQLrW9Xe11GFqP3SYA94Ptt1P |     4 | 150x150.jpg   | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmNydje7o5jmrJTxGEp8naviWsRjKwsXnJvVrZiiKYjewH |     5 | 480x480.jpg   | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmbbMKFxn7YKJGLVCPwFRmSxdQqyD1sMxxKF8Zei9Zanjr |     6 | 1000x1000.jpg | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUnyPBxMU3RKNB1f5xc8uMUYeJq3EuQ2zMnKtVijidrtW |     7 | original.jpg  | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUxeWCFAFCrBNniMzxDXSMRxzUNW8n65Qb7HggLnJiFGC |     8 |               | metadata
> (7 rows)
> ```
> 
> Confirm latest DB migration w/playlists has been applied:
> 
> ```shell
> ubuntu@hn-remote-dev-2:~/audius-protocol/creator-node$ ./node_modules/.bin/sequelize db:migrate:status
> 
> Sequelize CLI [Node: 14.17.5, CLI: 5.5.1, ORM: 4.44.4]
> 
> Loaded configuration file "sequelize/sequelize-config.json".
> Using environment "development".
> sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:245:13
> up 20181121184323-create-user.js
> up 20181122035259-create-file.js
> up 20181129193321-create-audius-user.js
> up 20181204021851-create-session-token.js
> up 20181207194524-create-track.js
> up 20181211183951-create-track-file.js
> up 20190711143402-add_uuid_to_users.js
> up 20190805032224-add_on_update_cascade_to_session_tokens.js
> up 20190827175430-add-types-to-files.js
> up 20190906221141-add-blocknumber-to-cnodeuser.js
> up 20191014204226-add-indexes.js
> up 20191108022324-associate-files-with-trackuuid.js
> up 20200623002237-add-parent-file-data.js
> up 20200724154816-sync-indexes.js
> up 20200904183206-add-track-uuid-index.js
> up 20200911004845-allow-track-and-audiusUsers-appends.js
> up 20200918150546-add-clock.js
> up 20200922131913-post_vector_clock_db_migrations.js
> up 20200928233308-create-content-blacklist.js
> up 20201105015155-add-cid-and-update-content-blacklist.js
> up 20210722020837-add-skipped-column-to-files-table.js
> up 20210831162548-add-index-files-sourcefile.js
> up 20220427014326-create-playlist-table-and-clock-record-type.js
> ```
> 
> Undo latest migration:
> 
> ```shell
> ubuntu@hn-remote-dev-2:~/audius-protocol/creator-node$ ./node_modules/.bin/sequelize db:migrate:undo --name 20220427014326-create-playlist-table-and-clock-record-type.js
> 
> Sequelize CLI [Node: 14.17.5, CLI: 5.5.1, ORM: 4.44.4]
> 
> Loaded configuration file "sequelize/sequelize-config.json".
> Using environment "development".
> sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:245:13
> == 20220427014326-create-playlist-table-and-clock-record-type: reverting =======
> == 20220427014326-create-playlist-table-and-clock-record-type: reverted (0.020s)
> ```
> 
> Next, updated cover photo and metadata without migration present:
> 
> ```shell
> audius_creator_node=# select * from "CNodeUsers" ;
>  lastLogin |              walletPublicKey               |         createdAt          |         updatedAt          |            cnodeUserUUID             | latestBlockNumber | clock
> -----------+--------------------------------------------+----------------------------+----------------------------+--------------------------------------+-------------------+-------
>            | 0xee73894f48454c478789284d960486e27ab9bcc1 | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:49:46.996+00 | d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |              1649 |    15
> (1 row)
> 
> audius_creator_node=# select "cnodeUserUUID","multihash","clock","fileName","type" from "Files";
>             cnodeUserUUID             |                   multihash                    | clock |   fileName    |   type
> --------------------------------------+------------------------------------------------+-------+---------------+----------
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmX3uGhDU9a8sC5LRceH1LMemXwmQqfaLQVmTT4rM2tHtZ |     1 |               | metadata
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUwGD453zzANwGYgmH9cNjBbkWDXJYMZXKmGNKChXHws3 |     3 |               | dir
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmWgv9ZFTUXhaWEDvm7f8KQLrW9Xe11GFqP3SYA94Ptt1P |     4 | 150x150.jpg   | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmNydje7o5jmrJTxGEp8naviWsRjKwsXnJvVrZiiKYjewH |     5 | 480x480.jpg   | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmbbMKFxn7YKJGLVCPwFRmSxdQqyD1sMxxKF8Zei9Zanjr |     6 | 1000x1000.jpg | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUnyPBxMU3RKNB1f5xc8uMUYeJq3EuQ2zMnKtVijidrtW |     7 | original.jpg  | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUxeWCFAFCrBNniMzxDXSMRxzUNW8n65Qb7HggLnJiFGC |     8 |               | metadata
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmfJmo3gLrZKUrRQuV1YSDt6FmP8M3KPETmAdUgAFBE94h |    10 |               | dir
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmWmEV7fh6TmGJKLuBoH3PiJ93PxJypADAxKWR8b9R1Hgb |    11 | 640x.jpg      | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmeWdQ9bjeNbWitgkgPnxddSShhxET2hWBRmsUucRtSP36 |    12 | 2000x.jpg     | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmUhPPAWVNH3P5FMpQUUueKx81YL7Eu57wBsxLgmCSLbP6 |    13 | original.jpg  | image
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd | QmXR2Lx1ceCDQ9otvTmjbYL552rEr1qrZHcc4AR1fXgF1m |    14 |               | metadata
> (12 rows)
> ```
> 
> Re-running migration (upward testing w/state):
> 
> ```shell
> ubuntu@hn-remote-dev-2:~/audius-protocol/creator-node$ ./node_modules/.bin/sequelize db:migrate
> 
> Sequelize CLI [Node: 14.17.5, CLI: 5.5.1, ORM: 4.44.4]
> 
> Loaded configuration file "sequelize/sequelize-config.json".
> Using environment "development".
> sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:245:13
> == 20220427014326-create-playlist-table-and-clock-record-type: migrating =======
> == 20220427014326-create-playlist-table-and-clock-record-type: migrated (0.028s)
> ```
> 
> Inserting a record into Playlists table (following this -
> 
> https://github.com/AudiusProject/audius-protocol/blob/6fef5644f572934c5ce21e67d97682ce2ca5ac00/creator-node/src/dbManager.js#L273
> 
> )
> ```shell
> audius_creator_node=# update "CNodeUsers" set clock = clock + 1 where text("cnodeUserUUID") = 'd3f8e8ea-5e33-41bc-b004-8fc6ca4626cd';
> UPDATE 1
> 
> audius_creator_node=# select * from "CNodeUsers";
>  lastLogin |              walletPublicKey               |         createdAt          |         updatedAt          |            cnodeUserUUID             | latestBlockNumber | clock
> -----------+--------------------------------------------+----------------------------+----------------------------+--------------------------------------+-------------------+-------
>            | 0xee73894f48454c478789284d960486e27ab9bcc1 | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:49:46.996+00 | d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |              1649 |    16
> 
> INSERT INTO CLOCKRECORDS:
> 
> audius_creator_node=# INSERT INTO "ClockRecords" ("cnodeUserUUID", "clock", "sourceTable", "createdAt", "updatedAt") VALUES('d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd', 16, 'Playlist', '2022-06-22 21:29:13.755+00', '2022-06-22 21:29:13.755+00');
> 
> audius_creator_node=# select * from "ClockRecords";
>             cnodeUserUUID             | clock | sourceTable |         createdAt          |         updatedAt
> --------------------------------------+-------+-------------+----------------------------+----------------------------
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |     1 | File        | 2022-06-22 21:29:14.129+00 | 2022-06-22 21:29:14.129+00
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |     2 | AudiusUser  | 2022-06-22 21:29:18.675+00 | 2022-06-22 21:29:18.675+00
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |     3 | File        | 2022-06-22 21:35:45.041+00 | 2022-06-22 21:35:45.041+00
> ...
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    12 | File        | 2022-06-22 21:49:38.927+00 | 2022-06-22 21:49:38.927+00
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    13 | File        | 2022-06-22 21:49:38.934+00 | 2022-06-22 21:49:38.934+00
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    14 | File        | 2022-06-22 21:49:39.721+00 | 2022-06-22 21:49:39.721+00
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    15 | AudiusUser  | 2022-06-22 21:49:46.984+00 | 2022-06-22 21:49:46.984+00
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    16 | Playlist    | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:29:13.755+00
> 
> INSERT INTO PLAYLISTS:
> audius_creator_node=# INSERT INTO "Playlists" ("cnodeUserUUID", "metadataFileUUID","metadataJSON","blockchainId","clock","createdAt","updatedAt") VALUES('d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd','1a3ee179-6e89-473f-924c-36fc6089bda6','{}',1000,16,'2022-06-22 21:29:13.755+00','2022-06-22 21:29:13.755+00');
> 
> audius_creator_node=# select * from "Playlists";
>             cnodeUserUUID             | clock |           metadataFileUUID           | metadataJSON | blockchainId | playlistImageFileUUID |         createdAt          |         updatedAt
> --------------------------------------+-------+--------------------------------------+--------------+--------------+-----------------------+----------------------------+----------------------------
>  d3f8e8ea-5e33-41bc-b004-8fc6ca4626cd |    16 | 1a3ee179-6e89-473f-924c-36fc6089bda6 | {}           |         1000 |                       | 2022-06-22 21:29:13.755+00 | 2022-06-22 21:29:13.755+00
> ```
> 
> Downgrading migration:
> 
> ```shell
> == 20220427014326-create-playlist-table-and-clock-record-type: reverting =======
> DatabaseError [SequelizeDatabaseError]: invalid input value for enum "enum_ClockRecords_sourceTable": "Playlist"
>     at Query.formatError (/home/ubuntu/audius-protocol/creator-node/node_modules/sequelize/lib/dialects/postgres/query.js:363:16)
>     at /home/ubuntu/audius-protocol/creator-node/node_modules/sequelize/lib/dialects/postgres/query.js:86:18
>     at tryCatcher (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/util.js:16:23)
>     at Promise._settlePromiseFromHandler (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:547:31)
>     at Promise._settlePromise (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:604:18)
>     at Promise._settlePromise0 (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:649:10)
>     at Promise._settlePromises (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/promise.js:725:18)
>     at _drainQueueStep (/home/ubuntu/audius-protocol/creator-node/node_modules/bluebird/js/release/async.js:93:12)
> ```

ah yeah this makes sense. its prob fine to just not drop the playlist enum? like realistically we never use the down migrations, and some of our migrations prob don't have working down migrations anyway

hareeshnagaraj commented:
> 
> ah yeah this makes sense. its prob fine to just not drop the playlist enum? like realistically we never use the down migrations, and some of our migrations prob don't have working down migrations anyway

RE: Offline discussion - going to keep the enum type in down migration but remove everything else. Should be safe, but will test as well.

SidSethi reviewed (approved):
🚀
